aten/src/ATen/native/cpu/SortingKernel.cpp

```diff
@@ -139,30 +143,82 @@ static void parallel_sort1d_kernel(
 }
 #endif
 
+template< class RandomIt, class Compare>
+static void inline parallel_sort( RandomIt first, RandomIt last, Compare comp, std::function<void(RandomIt, RandomIt, Compare)> sort_f ) {
+  int64_t n = last - first;
+  const int64_t chunk_size = std::max(at::divup(n, at::get_num_threads()), at::divup(n, at::divup(n, 512)));
+  const int64_t num_threads = at::divup(n, chunk_size);
+
+  at::parallel_for(
+    0, n, chunk_size, [&](int64_t begin, int64_t end) {
+      sort_f(first + begin, first + end, comp);
+    }
+  );
+
+  if (num_threads > 1) {
+    const int64_t merge_threads = (num_threads - ((int) chunk_size * num_threads != n)) / 2;
+    at::parallel_for(
+      0, chunk_size * 2 * merge_threads, chunk_size * 2, [&](int64_t begin, int64_t end) {
+        std::inplace_merge(first + begin, first + begin + chunk_size, first + end, comp);
+      }
+    );
+
+    for (int i=0; i < merge_threads - 1; i++) {
+      std::inplace_merge(first, first + chunk_size * 2 * (i + 1), first + chunk_size * 2 * (i + 2), comp);
+    }
```

Comment:
Merge can be done in parallel and in a hierarchical manner. Here it looks like the parallelization is done only on a single level.

Reply:
Actually, sounds like a good thing to have at some point -- a generalized parallel implementation of divide-and-conquer types of algorithms...

Reply:
yes, although the conquer part is quite specific to each algorithm